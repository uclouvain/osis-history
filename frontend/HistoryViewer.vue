<!--
  -
  -   OSIS stands for Open Student Information System. It's an application
  -   designed to manage the core business of higher education institutions,
  -   such as universities, faculties, institutes and professional schools.
  -   The core business involves the administration of students, teachers,
  -   courses, programs and so on.
  -
  -   Copyright (C) 2015-2021 UniversitÃ© catholique de Louvain (http://www.uclouvain.be)
  -
  -   This program is free software: you can redistribute it and/or modify
  -   it under the terms of the GNU General Public License as published by
  -   the Free Software Foundation, either version 3 of the License, or
  -   (at your option) any later version.
  -
  -   This program is distributed in the hope that it will be useful,
  -   but WITHOUT ANY WARRANTY; without even the implied warranty of
  -   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  -   GNU General Public License for more details.
  -
  -   A copy of this license - GNU General Public License - is available
  -   at the root of the source code of this program.  If not,
  -   see http://www.gnu.org/licenses/.
  -
  -->
<template>
  <div class="history-viewer">
    <div
        v-if="modes.length > 1"
        class="btn-group"
        data-toggle="buttons"
    >
      <template
          v-for="buttonMode in modes"
          :key="buttonMode"
      >
        <label
            v-if="buttonMode === 'table'"
            class="btn btn-default"
            :class="{active: mode === 'TableHistory'}"
            @click="mode = 'TableHistory'"
        >
          <input
              type="radio"
              name="mode"
          >
          <span class="fa fa-table" />
        </label>
        <label
            v-else-if="buttonMode === 'vertical'"
            class="btn btn-default"
            :class="{active: mode === 'TimelineHistory' && horizontal === false}"
            @click="mode = 'TimelineHistory'; horizontal = false;"
        >
          <input
              type="radio"
              name="mode"
          >
          <span class="fas fa-ellipsis-v" />
        </label>
        <label
            v-else-if="buttonMode === 'horizontal'"
            class="btn btn-default"
            :class="{active: mode === 'TimelineHistory' && horizontal === true}"
            @click="mode = 'TimelineHistory'; horizontal = true;"
        >
          <input
              type="radio"
              name="mode"
          >
          <span class="fa fa-ellipsis-h" />
        </label>
      </template>
    </div>
    <div class="viewport">
      <span
          v-if="loading"
          class="spinner"
      />
      <span
          v-else-if="error"
          class="text-danger"
      >
        {{ error }}
      </span>
      <component
          :is="mode"
          v-else
          :entries="entries"
          :horizontal="horizontal"
          :on-item-render="mode === 'TableHistory' ? onItemRenderTable : (horizontal ? onItemRenderHorizontalTimeline : onItemRenderVerticalTimeline)"
          :on-headers-render="onHeadersRenderTable"
          :on-history-empty="onHeadersRenderTable"
      />
    </div>
  </div>
</template>

<script lang="ts">
import TableHistory from './components/TableHistory.vue';
import TimelineHistory from './components/TimelineHistory.vue';
import {defineComponent} from "vue";
import type {PropType} from "vue";
import type {Entry} from "./interfaces";


export default defineComponent({
  name: 'HistoryViewer',
  components: {TableHistory, TimelineHistory},
  props: {
    url: {
      type: String,
      required: true,
    },
    tags: {
      type: String,
      default: '',
    },
    modes: {
      type: Array as PropType<string[]>,
      default: () => ['table', 'vertical', 'horizontal'],
    },
    defaultMode: {
      type: String,
      default: 'table',
    },
    onItemRenderTable: {
      type: Function,
      default: null,
    },
    onHeadersRenderTable: {
      type: Function,
      default: null,
    },
    onItemRenderHorizontalTimeline: {
      type: Function,
      default: null,
    },
    onItemRenderVerticalTimeline: {
      type: Function,
      default: null,
    },
    onHistoryEmptyRender: {
      type: Function,
      default: null,
    },
  },
  data() {
    let defaultMode: string = this.defaultMode;
    if (!this.modes.includes(this.defaultMode)) {
      defaultMode = this.modes[0];
    }
    return {
      mode: defaultMode === 'table' ? 'TableHistory' : 'TimelineHistory',
      horizontal: defaultMode === 'horizontal',
      entries: [] as Entry[],
      error: '',
      loading: true,
    };
  },
  mounted() {
    void this.loadEntries();
  },
  methods: {
    loadEntries: async function () {
      try {
        const response = await fetch(`${this.url}${this.tags ? '?tags=' + this.tags : ''}`);
        if (response.status === 200) {
          this.entries = (await response.json() as Entry[]);
        } else {
          this.error = response.statusText;
        }
      } catch (e) {
        this.error = (e as Error).message;
      }
      this.loading = false;
    },
  },
});
</script>

<style>
.history-viewer .viewport {
  margin-top: 1em;
  overflow-x: scroll;
}

.history-viewer .spinner {
  display: block;
  margin: 0 auto;
  border: 3px solid #286090;
  border-top-color: transparent;
  border-radius: 50%;
  width: 20px;
  height: 20px;
  animation: spin 1.5s linear infinite;
}
</style>
