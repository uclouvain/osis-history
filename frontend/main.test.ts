/*
 *
 * OSIS stands for Open Student Information System. It's an application
 * designed to manage the core business of higher education institutions,
 * such as universities, faculties, institutes and professional schools.
 * The core business involves the administration of students, teachers,
 * courses, programs and so on.
 *
 * Copyright (C) 2015-2023 UniversitÃ© catholique de Louvain (http://www.uclouvain.be)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * A copy of this license - GNU General Public License - is available
 * at the root of the source code of this program.  If not,
 * see http://www.gnu.org/licenses/.
 *
 */

import {expect, test, vi} from 'vitest';
/* eslint-disable vue/prefer-import-from-vue */
import * as exports from '@vue/runtime-dom';
import HistoryViewer from "./HistoryViewer.vue";
import fetchMock from "fetch-mock";
import {createApp} from "vue";


declare global {
  interface Window {
    customOnHeadersRenderTable: CallableFunction,
  }
}

// importing main, which has an export, can be done only once, so we put two tests in one
test('mount app with conversions', async () => {
  fetchMock.reset().mock('path:/api', []);
  window.customOnHeadersRenderTable = () => ``;
  document.body.innerHTML = `
  <div
    class="history-viewer"
    data-url="/api"
  ></div>
  <div
    class="history-viewer"
    data-url="/api"
    data-modes="table,horizontal"
    data-on-headers-render-table="customOnHeadersRenderTable"
  ></div>`;

  const spy = vi.spyOn(exports, 'createApp').mockImplementation(createApp);

  // Executes main file
  const main = await import('./main');
  expect(main).toHaveProperty('filterXssAndFormat');
  expect(document.body.innerHTML).toMatchSnapshot();

  const pElement = document.querySelectorAll('[data-v-app]');
  expect(pElement).toHaveLength(2);

  expect(spy).toBeCalledTimes(2);
  expect(spy).toHaveBeenCalledWith(HistoryViewer, {
    url: '/api',
  });
  expect(spy).toHaveBeenCalledWith(HistoryViewer, {
    url: '/api',
    modes: ["table", "horizontal"],
    onHeadersRenderTable: window.customOnHeadersRenderTable,
  });
});

