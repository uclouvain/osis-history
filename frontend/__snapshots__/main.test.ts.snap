// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mount app with conversions 1`] = `
"
  <div id=\\"app\\" class=\\"history-viewer\\" data-url=\\"/api\\" data-v-app=\\"\\"><div class=\\"history-viewer\\"><div class=\\"btn-group\\" data-toggle=\\"buttons\\"><label class=\\"btn btn-default active\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fa fa-table\\"></span></label><label class=\\"btn btn-default\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fas fa-ellipsis-v\\"></span></label><label class=\\"btn btn-default\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fa fa-ellipsis-h\\"></span></label></div><div class=\\"viewport\\"><table class=\\"table table-striped\\" horizontal=\\"false\\"><thead><tr><th>Date</th><th>Message</th><th>Author</th></tr></thead><tbody><tr><td colspan=\\"3\\">No history entry yet.</td></tr></tbody></table></div></div></div>
  <div id=\\"app1\\" class=\\"history-viewer\\" data-url=\\"/api\\" data-modes=\\"table,horizontal\\" data-on-headers-render-table=\\"customOnHeadersRenderTable\\" data-v-app=\\"\\"><div class=\\"history-viewer\\"><div class=\\"btn-group\\" data-toggle=\\"buttons\\"><label class=\\"btn btn-default active\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fa fa-table\\"></span></label><label class=\\"btn btn-default\\"><input type=\\"radio\\" name=\\"mode\\"><span class=\\"fa fa-ellipsis-h\\"></span></label></div><div class=\\"viewport\\"><table class=\\"table table-striped\\" horizontal=\\"false\\"><thead></thead><tbody><tr><td colspan=\\"3\\">No history entry yet.</td></tr></tbody></table></div></div></div>"
`;
